sql/parser/builtins.go

```diff
@@ -72,6 +74,75 @@ var builtins = map[string]builtin{
 	"upper": stringBuiltin(func(s string) (Datum, error) {
 		return DString(strings.ToUpper(s)), nil
 	}),
+
+	// TODO(XisiHuang): support the substring(str FROM x FOR y) syntax
+	"substring": {
+		nArgs: -1,
+		fn: func(args DTuple) (Datum, error) {
+			argsNum := len(args)
+			if argsNum != 2 && argsNum != 3 {
+				return DNull, fmt.Errorf("incorrect number of arguments: %d vs %s",
+					argsNum, "2 or 3")
+			}
+
+			dstr, ok := args[0].(DString)
+			if !ok {
+				return DNull, argTypeError(args[0], "string")
+			}
+			str := string(dstr)
+
+			start, ok := args[1].(DInt)
+			if !ok {
+				return DNull, argTypeError(args[1], "int")
+			}
+
+			s := int(start) - 1
+
+			var e int
+			if argsNum == 2 {
+				e = len(str)
+			} else {
+				count, ok := args[2].(DInt)
+				if !ok {
+					return DNull, argTypeError(args[2], "int")
+				}
+				c := int(count)
+				if c < 0 {
+					return DNull, invalidArgError(fmt.Sprintf("negative substring length not allowed: %d", c))
+				}
+				e = s + c
+			}
+
+			if e <= 0 || s > len(str)-1 {
```

Comment:
@tamird No, s could be negative, same as in postgres, so e may also be negative.